File: l33tcode/snapshot-array.py
from dataclasses import dataclass
import bisect

@dataclass
class Value:
    value: int
    snapshot_id: int

    def __eq__(self, other) -> bool:
        return self.snapshot_id == other.snapshot_id

    def __lt__(self, other) -> bool:
        return self.snapshot_id < other.snapshot_id

class SnapshotArray:
    _values: list[list[Value]]
    _snapshot_id: int

    def __init__(self, length: int):
        self._values = [[Value(0, 0)] for _ in range(length)]
        self._snapshot_id = 0

    def set(self, index: int, val: int) -> None:
        if self._values[index][-1].snapshot_id == self._snapshot_id:
            self._values[index][-1] = Value(val, self._snapshot_id)
        else:
            self._values[index].append(Value(val, self._snapshot_id))

    def snap(self) -> int:
        self._snapshot_id += 1

        return self._snapshot_id - 1

    def get(self, index: int, snap_id: int) -> int:
        pos = bisect.bisect_right(self._values[index], Value(0, snap_id))

        return self._values[index][pos-1].value

File: l33tcode/snapshot-array.go
package main

type value struct {
	value      int
	snapshotID int
}

type SnapshotArray struct {
	values     [][]value
	snapshotID int
}

func Constructor(length int) SnapshotArray {
	values := make([][]value, length)
	for i := range values {
		values[i] = []value{{}}
	}
	return SnapshotArray{
		values: values,
	}
}

func (this *SnapshotArray) Set(index int, val int) {
	if this.values[index][len(this.values[index])-1].snapshotID != this.snapshotID {
		this.values[index] = append(this.values[index], value{})
	}
	this.values[index][len(this.values[index])-1] = value{val, this.snapshotID}
}

func (this *SnapshotArray) Snap() int {
	this.snapshotID += 1
	return this.snapshotID - 1
}

func (this *SnapshotArray) Get(index int, snap_id int) int {
	left, right := 0, len(this.values[index])

	for left < right {
		mid := left + (right-left)/2

		if this.values[index][mid].snapshotID <= snap_id {
			left = mid + 1
		} else {
			right = mid
		}
	}

	return this.values[index][left-1].value
}
